nixos/modules/services/matrix/synapse.nix

```diff
@@ -151,10 +169,53 @@ in {
 
       package = mkOption {
         type = types.package;
-        default = pkgs.matrix-synapse;
-        defaultText = literalExpression "pkgs.matrix-synapse";
+        readOnly = true;
+        description = lib.mdDoc ''
+          Reference to the `matrix-synapse` wrapper with all extras
+          (e.g. for `oidc` or `saml2`) added to the `PYTHONPATH` of all executables.
+
+          This option is useful to reference the "final" `matrix-synapse` package that's
+          actually used by `matrix-synapse.service`. For instance, when using
+          workers, it's possible to run
+          `''${config.services.matrix-synapse.package}/bin/synapse_worker` and
+          no additional PYTHONPATH needs to be specified for extras or plugins configured
+          via `services.matrix-synapse`.
+
+          However, this means that this option is supposed to be only declared
+          by the `services.matrix-synapse` module itself and is thus read-only.
+          In order to modify `matrix-synapse` itself, use an overlay to override
+          `pkgs.matrix-synapse-unwrapped`.
```

Comment:
Should we have a `package-unwrapped` option or similar?

Reply:
Writing an overlay feels like a lot more work than setting an option. I don't feel too strongly, though, since I don't see myself ever needing this, short of manually deploying a hotfix or something.

Reply:
Is `package = pkgs.matrix-synapse-unwrapped.overrideAttrs` that much more than `self: super: { matrix-synapse-unwrapped = super.matrix-synapse-unwrapped.overrideAttrs` though?

Reply:
This is rather weird and unaligned with most other modules. Also stacking overlays is more computetional expensive than another option and it breaks people's configs which have overwritten the package in the past.

Reply:
> This is rather weird and unaligned with most other modules

This `package`-option isn't even standardized to my knowledge and in that case I considered the ability to access the "final" synapse package trivially more important than having _another_ override mechanism.

> Also stacking overlays is more computetional expensive than another option and it breaks people's configs which have overwritten the package in the past.

Do you have any numbers on that? I'm somewhat skeptical about this claim (and haven't seen any benchmark/blogpost/?) on that topic because of Nix's lazy nature the fixpoint computation shouldn't be too expensive. Or is this a direct comparison to `package`?

Also, overlays make it at least possible to combine several changes to a single package on evaluation (i.e. >1 overlay). When you try to do this with `package =`, then it can't be merged by `types.package`.